qiskit_nature/circuit/library/initial_states/hartree_fock.py

```diff
@@ -26,49 +29,47 @@ class HartreeFock(QuantumCircuit):
     """A Hartree-Fock initial state."""
 
     def __init__(self,
-                 num_orbitals: int,
+                 num_spin_orbitals: int,
                  num_particles: Union[Tuple[int, int], int],
-                 qubit_mapping: str = 'parity',
-                 two_qubit_reduction: bool = True,
-                 sq_list: Optional[List[int]] = None) -> None:
+                 qubit_converter: QubitConverter) -> None:
         """
         Args:
-            num_orbitals: The number of spin orbitals, has a min. value of 1.
+            num_spin_orbitals: The number of spin orbitals, has a min. value of 1.
             num_particles: The number of particles. If this is an integer, it is the total (even)
                 number of particles. If a tuple, the first number is alpha and the second number is
                 beta.
-            qubit_mapping: The mapping type for qubit operator.
-            two_qubit_reduction: A flag indicating whether or not two qubit is reduced.
-            sq_list: The position of the single-qubit operators that anticommute with the cliffords.
+            qubit_converter: a QubitConverter instance.
         """
+
         # get the bitstring encoding the Hartree Fock state
-        bitstr = hartree_fock_bitstring(num_orbitals, num_particles, qubit_mapping,
-                                        two_qubit_reduction, sq_list)
+        bitstr = hartree_fock_bitstring(num_spin_orbitals, num_particles)
+
+        # encode the bitstring as a `FermionicOp`
+        label = ['+' if bit else 'I' for bit in bitstr]
+        bitstr_op = SecondQuantizedOp([FermionicOp(''.join(label))])
+
+        # map the `FermionicOp` to a qubit operator
+        qubit_op = qubit_converter.to_qubit_ops([bitstr_op])[0]
 
         # construct the circuit
-        qr = QuantumRegister(len(bitstr), 'q')
+        qr = QuantumRegister(qubit_op.num_qubits, 'q')
         super().__init__(qr, name='HF')
 
-        # add gates in the right positions
-        for i, bit in enumerate(reversed(bitstr)):
+        # Add gates in the right positions: we are only interested in the `X` gates because we want
+        # to create particles (0 -> 1) where the initial state introduced a creation (`+`) operator.
+        for i, bit in enumerate(qubit_op.primitive.table.X[0]):
             if bit:
                 self.x(i)
 
 
-def hartree_fock_bitstring(num_orbitals: int,
-                           num_particles: Union[Tuple[int, int], int],
-                           qubit_mapping: str = 'parity',
-                           two_qubit_reduction: bool = True,
-                           sq_list: Optional[List[int]] = None) -> List[bool]:
+def hartree_fock_bitstring(num_spin_orbitals: int,
```

Comment:
What is the thinking here tapering this bitstring accordingly? The qubit_convertor has the z2_symmetries getter that would pass back a Z2Symmetry object with the sq_list in it. I imagine we will add the two_qubit_reduction to the Qubit_Convertor too, as a flag, so it can do that when the mapper supports it (the mapper has already had the flag added to it). The reason this method was public was so that it could be use by the stack, maybe the algorithms now, as input to determine the desired sector for ground state, and as such, since two qubit reduction  is done ahead of symmetry reduction the sqlist in that case would be empty. Of course by the time HF would be used to create the state then the symettry would be done and the sq_list (from Z2Symmetry) available

Reply:
I saw only later that this method was used elsewhere in the stack. I am not sure to what degree this will still be necessary though.
Originally, I had planned to make this method return the fully mapped and tapered `QubitOperator`. It would then also take the `QubitConverter` as an input. Is that what you mean here?
I am not sure from the top of my head, if we can then extract the correct sector from the final result. Is that stored in the `QubitOperator` somehow? Or will we make the `QubitConverter` stateful and store it in there?

Reply:
The sector with the ground state, that we be chosen, is determined using that HF string. I think there is a pick/select sector. At the time we need to do symmetry reduction the operator will be mapped, and have had any two qubit reduction done. So the string needed to reflect that. I think the sqlist coming from the symmetry is used later in conjunction with UCCSD since by that time its working on the completly mapped/symmeterized operator.

Since QubitConvertor needs to know which sector to select when tapering, I was imagining the algorithm/problem knowing it was the ground state being computed that it would use the pick/select sector, using the HF info, and pass that in when calling the to_qubit_op(). So as mentioned, to taper it, it would just have the mapping and two q reduction flag settings to determine the sector from the mapped/2 q reduced operator.

This is what the code that did auto sector determination looked like before https://github.com/Qiskit/qiskit-aqua/blob/39067b4f9d345ca0fb2eeab41b231e263f828209/qiskit/chemistry/core/hamiltonian.py#L305-L313 The pick sector is further down

Yes, there is TaperedPauliSumOp (as a sub-class of PauliSumOp) that can carry the Z2Symmetry; but I think any operations on it will lose that information, as its a very simple sub class that does not override any methods to preserve that where it made sense - such as simplification, multiplication by constant  etc

Reply:
After our offline discussion, I have added a TODO comment documenting the currently open design questions.
However, as decided offline, let's postpone the resolution of this TODO until we finalize the `QubitConverter` implementation and know the details of its API and integration into the workflow.